cirq/google/optimizers/optimize_for_sycamore.py

```diff
@@ -12,73 +12,113 @@
+        optimizers.MergeInteractions(tolerance=tolerance,
+                                     allow_partial_czs=True).optimize_circuit,
+        lambda c: optimizers.merge_single_qubit_gates_into_phased_x_z(
+            c, tolerance),
+        optimizers.EjectPhasedPaulis(tolerance=tolerance).optimize_circuit,
+        optimizers.EjectZ(tolerance=tolerance).optimize_circuit,
+        optimizers.DropNegligible(tolerance=tolerance).optimize_circuit,
+    ]
+
+
+def _get_sycamore_optimizers(tolerance: float, tabulation: Optional[GateTabulation]) \
+        -> List[Callable[['cirq.Circuit'], None]]:
+    return [
+        ConvertToSycamoreGates(tabulation=tabulation).optimize_circuit,
+        lambda c: optimizers.merge_single_qubit_gates_into_phased_x_z(
+            c, tolerance),
+        optimizers.EjectPhasedPaulis(tolerance=tolerance).optimize_circuit,
+        optimizers.EjectZ(tolerance=tolerance).optimize_circuit,
+        optimizers.DropNegligible(tolerance=tolerance).optimize_circuit,
+    ]
+
+
+def _get_sqrt_iswap_optimizers(tolerance: float, tabulation: Optional[GateTabulation]) \
+        -> List[Callable[['cirq.Circuit'], None]]:
+    if tabulation is not None:
+        raise ValueError("Gate tabulation not supported for sqrt_iswap")
+    return [
+        ConvertToSqrtIswapGates().optimize_circuit,
+        lambda c: optimizers.merge_single_qubit_gates_into_phased_x_z(
+            c, tolerance),
+        optimizers.EjectPhasedPaulis(tolerance=tolerance).optimize_circuit,
+        optimizers.EjectZ(tolerance=tolerance).optimize_circuit,
+        optimizers.DropNegligible(tolerance=tolerance).optimize_circuit,
+    ]
+
 
 _OPTIMIZER_TYPES = {
-    'xmon': _XMON_OPTIMIZERS,
-    'xmon_partial_cz': _XMON_OPTIMIZERS_PART_CZ,
-    'sqrt_iswap': _SQRT_ISWAP_OPTIMIZERS,
-    'sycamore': _SYCAMORE_OPTIMIZERS,
+    'xmon': _get_xmon_optimizers,
+    'xmon_partial_cz': _get_xmon_optimizers_part_cz,
+    'sqrt_iswap': _get_sqrt_iswap_optimizers,
+    'sycamore': _get_sycamore_optimizers,
 }
 
 
+@lru_cache()
+def _gate_product_tabulation_cached(optimizer_type, tabulation_resolution):
+    random_state = np.random.RandomState(51)
+    if optimizer_type == 'sycamore':
+        return gate_product_tabulation(protocols.unitary(cg_ops.SYC),
+                                       tabulation_resolution,
+                                       random_state=random_state)
+    else:
+        raise NotImplementedError(
+            f"Gate tabulation not supported for {optimizer_type}")
+
+
 def optimized_for_sycamore(
         circuit: 'cirq.Circuit',
         new_device: Optional['cirq.google.XmonDevice'] = None,
         qubit_map: Callable[['cirq.Qid'], devices.GridQubit] = lambda e: cast(
             devices.GridQubit, e),
-        optimizer_type: str = 'sqrt_iswap') -> 'cirq.Circuit':
+        optimizer_type: str = 'sqrt_iswap',
+        tolerance: float = 1e-5,
+        tabulation_resolution: Optional[float] = None,
```

Comment:
We should force keywords starting from `new_device`.

Reply:
done